packages/devtools_app/lib/src/memory/memory_timeline.dart

```diff
@@ -0,0 +1,483 @@
+  }
+
+  void resume() {
+    manuallyPaused = false;
+    _pausedNotifier.value = false;
+  }
+
+  /// Notifies any visible marker for a particular chart should be hidden.
+  final _markerHiddenNotifier = ValueNotifier<bool>(false);
+
+  ValueListenable<bool> get markerHiddenNotifier => _markerHiddenNotifier;
+
+  void hideMarkers() {
+    _markerHiddenNotifier.value = !_markerHiddenNotifier.value;
+  }
+
+  /// dart_ui.Image Image asset displayed for each entry plotted in a chart.
+  // ignore: unused_field
+  dart_ui.Image _img;
+
+  // TODO(terry): Look at using _img for each data point (at least last N).
+  static dart_ui.Image get dataPointImage => null;
+
+  set image(dart_ui.Image img) {
+    _img = img;
+  }
+
+  void reset() {
+    liveData.clear();
+    startingIndex = 0;
+    dartChartData.reset();
+    eventsChartData.reset();
+    androidChartData.reset();
+  }
+
+  /// Y-coordinate of an Event entry to not display (the no event state).
+  static const emptyEvent = -2.0; // Event (empty) to not visibly display.
+
+  /// Event to display in the event pane (User initiated GC, snapshot,
+  /// automatic snapshot, etc.)
+  static const visibleEvent = 1.7;
+
+  /// Monitor events Y axis.
+  static const visibleMonitorEvent = 0.85;
+
+  /// VM's GCs are displayed in a smaller glyph and closer to the heap graph.
+  static const visibleVmEvent = 0.2;
+
+  /// Common utility function to handle loading of the data into the
+  /// chart for either offline or live Feed.
+  List<Map> _processData(int index) {
+    final result = <Map<String, Entry>>[];
+
+    for (var lastIndex = index; lastIndex < data.length; lastIndex++) {
+      final sample = data[lastIndex];
+      final timestamp = sample.timestamp.toDouble();
+
+      // Flutter Framework memory (Dart VM Heaps)
+      final capacity = sample.capacity.toDouble();
+      final used = sample.used.toDouble();
+      final external = sample.external.toDouble();
+
+      // TOOD(terry): Need to plot.
+      final rss = (sample.rss ?? 0).toDouble();
+
+      final extEntry = Entry(x: timestamp, y: external, icon: dataPointImage);
+      final usedEntry =
+          Entry(x: timestamp, y: used + external, icon: dataPointImage);
+      final capacityEntry =
+          Entry(x: timestamp, y: capacity, icon: dataPointImage);
+      final rssEntry = Entry(x: timestamp, y: rss, icon: dataPointImage);
+
+      // Engine memory values (ADB Android):
+      final javaHeap = sample.adbMemoryInfo.javaHeap.toDouble();
+      final nativeHeap = sample.adbMemoryInfo.nativeHeap.toDouble();
+      final code = sample.adbMemoryInfo.code.toDouble();
+      final stack = sample.adbMemoryInfo.stack.toDouble();
+      final graphics = sample.adbMemoryInfo.graphics.toDouble();
+      final other = sample.adbMemoryInfo.other.toDouble();
+      final system = sample.adbMemoryInfo.system.toDouble();
+      final total = sample.adbMemoryInfo.total.toDouble();
+
+      final graphicsEntry = Entry(
```

Comment:
track a y and use `+=` to simplify the code and make it easier to re-arange the order in the future.

Reply:
Done.